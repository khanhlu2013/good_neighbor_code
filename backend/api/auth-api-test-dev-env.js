const route = require("express").Router();

const keys = require("../configs/keys");
const { FRONTEND_URL } = keys;

if (keys.NODE_ENV === "production") {
  throw Error("Unexpect production env");
}

route.get("/dummy_login_for_test_purpose", (req, res, next) => {
  const User = require("../models/user");
  const { email, name } = req.query;

  (async () => {
    const user = await User.findOneOrCreate(email, name);
    req.login(user, err => {
      if (err) {
        console.log(err);
        res.status(500).send();
      } else {
        return res.redirect(FRONTEND_URL);
      }
    });
  })().catch(next);
});

module.exports = route;
