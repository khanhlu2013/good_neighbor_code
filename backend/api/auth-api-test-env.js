const route = require('express').Router();

const keys = require('../configs/keys');
const {FRONTEND_URL} = keys;

if(keys.NODE_ENV !== 'test'){
  throw Error(`Unexpected ${keys.NODE_ENV} env`);
}

route.get('/dummy_login_for_test_purpose',async (req,res)=>{
  const User =  require('../models/user');
  const {email,name} = req.query;
  try{
    const user = await User.findOneOrCreate(email,name);
    req.login(user,(err)=>{
      if(err){
        console.log(err);
        res.status(500).send();
      }else{
        return res.redirect(FRONTEND_URL);
      }
    });
  }catch(e){
    console.log(e);
    res.status(500).send();
  }
});

module.exports = route;