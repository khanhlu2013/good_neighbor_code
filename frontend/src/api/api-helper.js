import { Connection } from "../model/connection";
import { User } from "../model/user";
import { Post } from "../model/post";
import { Share } from "../model/share";

const rawToConnection = raw => {
  const {
    _id: id,
    from: { _id: fromID, email: fromEmail, name: fromName },
    to: { _id: toID, email: toEmail, name: toName },
    approvedByTo,
    approvedByFrom
  } = raw;

  const from = new User(fromID, fromEmail, fromName);
  const to = new User(toID, toEmail, toName);
  return new Connection(id, from, to, approvedByTo, approvedByFrom);
};

const rawsToConnections = raws => {
  return raws.map(raw => rawToConnection(raw));
};

function rawsToPosts(raws) {
  return raws.map(raw => {
    const { post: postRaw, user: userRaw } = raw;
    const sharesRaw = raw.shares.filter(share => share._id !== undefined);

    const shares = sharesRaw.map(shareRaw => {
      const { _id: userId, email, name } = shareRaw.borrower;
      const borrower = new User(userId, email, name);
      return new Share(
        shareRaw._id,
        borrower,
        shareRaw.dateCreated,
        shareRaw.isApprovedByFrom,
        shareRaw.isReturnedByTo,
        null //post to be set later
      );
    });
    const { _id: userId, email: userEmail, name: userName } = userRaw;
    const post = new Post(
      postRaw._id,
      new User(userId, userEmail, userName),
      postRaw.isActive,
      postRaw.title,
      postRaw.description,
      postRaw.dateCreated,
      shares
    );
    for (const share of post.shares) {
      share.setPost(post);
    }
    return post;
  });
}

export { rawsToConnections, rawToConnection, rawsToPosts };
