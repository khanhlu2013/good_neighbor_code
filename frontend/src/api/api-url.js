import RouteNode from "route-node";

const { NODE_ENV } = process.env;
let BACKEND_BASE_URL;
let BACKEND_PORT;

if (NODE_ENV === "production") {
  BACKEND_BASE_URL = "https://goodneighbor-backend-test.herokuapp.com";
  BACKEND_PORT = "";
} else {
  BACKEND_BASE_URL = "http://localhost";
  if (process.env.REACT_APP_NODE_ENV === "test") {
    BACKEND_PORT = ":3001";
  } else {
    BACKEND_PORT = ":3000";
  }
}

const BACKEND_URL = `${BACKEND_BASE_URL}${BACKEND_PORT}`;

const API_NODE = new RouteNode("", "", [
  new RouteNode("profile", "/profile", [
    { name: "authCheck", path: "/authCheck" },
    { name: "logout", path: "/logout" },
    { name: "searchEmail", path: "/searchEmail?:email" },
    // - connection
    { name: "connections", path: "/connections" },
    { name: "createConnection", path: "/createConnection" },
    { name: "updateConnection", path: "/updateConnection" },
    // - post
    { name: "createPost", path: "/createPost" },
    { name: "updatePost", path: "/updatePost" },
    { name: "awareReturnPost", path: "/awareReturnPost" },
    // - outPosts
    { name: "outPosts", path: "/outPosts" },
    // - inPosts
    { name: "inPosts", path: "/inPosts" },
    // - share
    { name: "share", path: "/share" },
    { name: "createShare", path: "/createShare" },
    { name: "deleteShare", path: "/deleteShare" },
    { name: "approveShare", path: "/approveShare" },
    { name: "returnShare", path: "/returnShare" },
    { name: "awareApproveShare", path: "/awareApproveShare" }
  ]),
  new RouteNode("auth", "/auth", [
    { name: "google", path: "/google" },
    {
      name: "backdoorLogin",
      path: "/backdoorLogin?:email?:name"
    }
  ])
]);

const API_URL = (dottedPath, params) => {
  const path = API_NODE.buildPath(dottedPath, params);
  if (!path) {
    throw Error(`${dottedPath} is not found!`);
  }

  return BACKEND_URL + path;
};

export { API_URL };
