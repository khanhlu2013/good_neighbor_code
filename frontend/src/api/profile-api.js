import keys from "../configs/keys.js";

const profile = async () => {
  const request = await fetch(keys.API_URL("profile"), {
    credentials: "include"
  });
  if (request.status === 401) {
    return null;
  }
  return await request.json();
};

const searchEmail = email => {
  return get("profile.searchEmail", { email });
};

const connections = () => {
  return get("profile.connections", {});
};

const modifyConnection = (connectionId, isApproved) => {
  return post("profile.modifyConnection", {
    connectionId,
    isApproved
  });
};

const createConnection = userIdToAdd => {
  return post("profile.createConnection", {
    userIdToAdd
  });
};

const posts = async () => {
  return get("profile.posts", {});
};

const crudPost = async (postID, title, description) => {
  const crudedPost = await post("profile.crudPost", {
    postID,
    title,
    description
  });
  return crudedPost;
};

//- helper ----

async function getJSON(response) {
  const text = await response.text();
  if (!text) {
    return null;
  }
  return JSON.parse(text);
}

async function post(dottedPath, params) {
  const response = await fetch(keys.API_URL(dottedPath), {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(params),
    credentials: "include"
  });

  return await getJSON(response);
}

async function get(dottedPath, params) {
  const url = keys.API_URL(dottedPath, params);
  const response = await fetch(url, { credentials: "include" });
  return await getJSON(response);
}

//-------------

const API = {
  profile,
  searchEmail,
  connections,
  modifyConnection,
  createConnection,
  posts,
  crudPost
};
export { API };
