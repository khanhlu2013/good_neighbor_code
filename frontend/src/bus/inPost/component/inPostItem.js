import React from "react";
import PropTypes from "prop-types";

import PostItemBody from "../../post/component/postItem_body";
import InPostItemFoot from "./foot/inPostItem_foot";
import PostItemStyle from "../../post/component/style/postItem_style";
import InPostItemHead from "./inPostItem_head";

function InPostItem(props) {
  const {
    loginUser,
    post,
    isRequestingPost,
    isDeleteingShare,
    isAwaringShare,
    isReturningShare,
    onRequestPost,
    onUnRequestPost,
    onAwareApprovePost,
    onReturnPost
  } = props;
  return (
    <PostItemStyle id="inPost-item-react">
      <InPostItemHead postUser={post.user} dateCreate={post.dateCreate} />
      <PostItemBody post={post} />
      <InPostItemFoot
        postId={post.id}
        loginUser={loginUser}
        isActive={post.isActive}
        currentlyBorrowShare={post.curBorrowShare}
        myRequestShare={
          post.requestShares.find(
            share => share.borrower.id === loginUser.id
          ) || null
        }
        isRequestingPost={isRequestingPost}
        isDeleteingShare={isDeleteingShare}
        isAwaringShare={isAwaringShare}
        isReturningShare={isReturningShare}
        onRequestPost={onRequestPost}
        onUnRequestPost={onUnRequestPost}
        onAwareApprovePost={onAwareApprovePost}
        onReturnPost={onReturnPost}
      />
    </PostItemStyle>
  );
}
InPostItem.propTypes = {
  loginUser: PropTypes.object.isRequired,
  post: PropTypes.object.isRequired,
  isRequestingPost: PropTypes.bool.isRequired,
  isDeleteingShare: PropTypes.bool.isRequired,
  isAwaringShare: PropTypes.bool.isRequired,
  isReturningShare: PropTypes.bool.isRequired,
  onRequestPost: PropTypes.func.isRequired,
  onUnRequestPost: PropTypes.func.isRequired,
  onAwareApprovePost: PropTypes.func.isRequired,
  onReturnPost: PropTypes.func.isRequired
};

export { InPostItem };
