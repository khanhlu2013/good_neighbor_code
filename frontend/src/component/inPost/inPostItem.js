import React from "react";
import PropTypes from "prop-types";
import { InPostItemFooting } from "./inPostItem_footing";
import { PostItemBody } from "../post/postItemBody";
import { InPostItemHead } from "./inPostItemHead";

function InPostItem(props) {
  const {
    loginUser,
    post,
    isRequestingPost,
    isDeleteingShare,
    isAwaringShare,
    isReturningShare,
    onCreateShare,
    onDeleteShare,
    onAwareShare,
    onReturnShare
  } = props;
  return (
    <div id="inPost-item-react" className="post-item shadow-box">
      <InPostItemHead postUser={post.user} dateCreate={post.dateCreate} />
      <PostItemBody post={post} />
      <InPostItemFooting
        postId={post.id}
        loginUser={loginUser}
        isActive={post.isActive}
        curBorrowShare={post.curBorrowShare}
        myRequestShare={
          post.requestShares.find(
            share => share.borrower.id === loginUser.id
          ) || null
        }
        isRequestingPost={isRequestingPost}
        isDeleteingShare={isDeleteingShare}
        isAwaringShare={isAwaringShare}
        isReturningShare={isReturningShare}
        onCreateShare={onCreateShare}
        onDeleteShare={onDeleteShare}
        onAwareShare={onAwareShare}
        onReturnShare={onReturnShare}
      />
    </div>
  );
}
InPostItem.propTypes = {
  loginUser: PropTypes.object.isRequired,
  post: PropTypes.object.isRequired,
  isRequestingPost: PropTypes.bool.isRequired,
  isDeleteingShare: PropTypes.bool.isRequired,
  isAwaringShare: PropTypes.bool.isRequired,
  isReturningShare: PropTypes.bool.isRequired,
  onCreateShare: PropTypes.func.isRequired,
  onDeleteShare: PropTypes.func.isRequired,
  onAwareShare: PropTypes.func.isRequired,
  onReturnShare: PropTypes.func.isRequired
};

export { InPostItem };
