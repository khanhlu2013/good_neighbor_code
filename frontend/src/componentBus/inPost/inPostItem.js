import React from "react";
import PropTypes from "prop-types";

import { InPostItemFoot } from "./inPostItem_foot";
import { PostItemBody } from "../post/postItem_body";
import { InPostItemHead } from "./inPostItem_head";
import { PostItemStyle } from "../post/style/postItem_style";

function InPostItem(props) {
  const {
    loginUser,
    post,
    isRequestingPost,
    isDeleteingShare,
    isAwaringShare,
    isReturningShare,
    onRequestPost,
    onUnRequestPost,
    onAwareShare,
    onReturnShare
  } = props;
  return (
    <PostItemStyle id="inPost-item-react">
      <InPostItemHead postUser={post.user} dateCreate={post.dateCreate} />
      <PostItemBody post={post} />
      <InPostItemFoot
        postId={post.id}
        loginUser={loginUser}
        isActive={post.isActive}
        curBorrowShare={post.curBorrowShare}
        myRequestShare={
          post.requestShares.find(
            share => share.borrower.id === loginUser.id
          ) || null
        }
        isRequestingPost={isRequestingPost}
        isDeleteingShare={isDeleteingShare}
        isAwaringShare={isAwaringShare}
        isReturningShare={isReturningShare}
        onRequestPost={onRequestPost}
        onUnRequestPost={onUnRequestPost}
        onAwareShare={onAwareShare}
        onReturnShare={onReturnShare}
      />
    </PostItemStyle>
  );
}
InPostItem.propTypes = {
  loginUser: PropTypes.object.isRequired,
  post: PropTypes.object.isRequired,
  isRequestingPost: PropTypes.bool.isRequired,
  isDeleteingShare: PropTypes.bool.isRequired,
  isAwaringShare: PropTypes.bool.isRequired,
  isReturningShare: PropTypes.bool.isRequired,
  onRequestPost: PropTypes.func.isRequired,
  onUnRequestPost: PropTypes.func.isRequired,
  onAwareShare: PropTypes.func.isRequired,
  onReturnShare: PropTypes.func.isRequired
};

export { InPostItem };
