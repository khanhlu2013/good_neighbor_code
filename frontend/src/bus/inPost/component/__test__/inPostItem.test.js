import React from "react";
import { shallow } from "enzyme";
import { rawsToPosts } from "../../../../api/_private_api_helper";
import { InPostItem } from "../inPostItem";

describe("InPostItem", () => {
  it("when render -> can match snapshot", () => {
    const fixtures = [
      {
        shares: [{}],
        user: {
          _id: "5bf2130ed95b19000076af71",
          email: "my@friend.com",
          name: "My Friend"
        },
        post: {
          _id: "5bf213205407a8bb7557848d",
          isActive: true,
          user: "5bf2130ed95b19000076af71",
          title: "post title",
          description: "post description",
          dateCreate: "2018-11-19T01:34:24.839Z",
          __v: 0
        }
      }
    ];
    const [post] = rawsToPosts(fixtures);
    const wrap = shallow(<InPostItem post={post} />);
    expect(wrap).toMatchSnapshot();
  });
});
