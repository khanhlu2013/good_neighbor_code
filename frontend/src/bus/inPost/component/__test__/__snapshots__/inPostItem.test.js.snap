// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InPostItem when render -> can match snapshot 1`] = `
<styled.div
  id="inPost-item-react"
>
  <InPostItemHead
    dateCreate={2018-11-19T01:34:24.839Z}
    postUserEmail="my@friend.com"
    postUserName="My Friend"
  />
  <PostItemBody
    post={
      Post {
        "dateCreate": 2018-11-19T01:34:24.839Z,
        "description": "post description",
        "id": "5bf213205407a8bb7557848d",
        "isActive": true,
        "shares": Array [],
        "title": "post title",
        "user": User {
          "email": "my@friend.com",
          "id": "5bf2130ed95b19000076af71",
          "name": "My Friend",
          "profileImageUrl": undefined,
        },
      }
    }
  />
  <Connect(InPostItemFoot)
    postId="5bf213205407a8bb7557848d"
  />
</styled.div>
`;
